docs/tutorials/gradient-checkpointing.md

```diff
@@ -0,0 +1,488 @@
+
+```{code-cell}
+def chain_compose(funs):
+  def f(x):
+    for fun in funs:
+      x = fun(x)
+    return x
+  return f
+
+f = chain_compose([jnp.sin] * 8)
+print_saved_residuals(f, 3.)
+```
+
+In general, the number of stored residuals scales linearly with the length of the chain:
+
+```{code-cell}
+f = chain_compose([jnp.sin] * 16)
+print_saved_residuals(f, 3.)
+```
+
+But you can apply {func}`jax.checkpoint` recursively to improve the scaling:
+
+```{code-cell}
+def recursive_checkpoint(funs):
+  if len(funs) == 1:
+    return funs[0]
+  elif len(funs) == 2:
+    f1, f2 = funs
+    return lambda x: f1(f2(x))
+  else:
+    f1 = recursive_checkpoint(funs[:len(funs)//2])
+    f2 = recursive_checkpoint(funs[len(funs)//2:])
+    return lambda x: f1(jax.checkpoint(f2)(x))
+```
+
+```{code-cell}
+f = recursive_checkpoint([jnp.sin] * 8)
+print_saved_residuals(f, 3.)
+```
+
+```{code-cell}
+f = recursive_checkpoint([jnp.sin] * 16)
+print_saved_residuals(f, 3.)
+```
+
+The cost here, as usual, is recomputation: in particular, you end up performing $\mathcal{O}(\log_2 D)$ times as many FLOPs:
+
+```{code-cell}
+f = chain_compose([jnp.sin] * 8)
+print_fwd_bwd(f, 3.)
+```
+
+```{code-cell}
+f = recursive_checkpoint([jnp.sin] * 8)
+print_fwd_bwd(f, 3.)
+```
+
+(gradient-checkpointing-practical-notes)=
+### Practical notes
+
+When differentiated functions are staged out to XLA for compilation — for example by applying {func}`jax.jit` to a function which contains a {func}`jax.grad` call — XLA will automatically optimize the computation, including decisions about when to compute or rematerialize values. As a result, **{func}`jax.checkpoint` often isn't needed for differentiated functions under a {func}`jax.jit`**. XLA will optimize things for you.
+
+One exception is when using staged-out control flow, like {func}`jax.lax.scan`. Automatic compiler optimizations across multiple control flow primitives (for example, across a forward-pass `scan` and the corresponding backward-pass `scan`), typically aren't aren't as thorough. As a result, it's often a good idea to use {func}`jax.checkpoint` on the body function passed to {func}`jax.lax.scan`.
+
+For example, one common pattern in large [Transformer models](https://en.wikipedia.org/wiki/Transformer_(machine_learning_model)) is to express the architecture as a {func}`jax.lax.scan` over layers so as to reduce compilation times. That is, using a simple fully-connected network as an analogy, instead of writing something like this:
+
+```python
+LayerParam = tuple[jnp.ndarray, jnp.ndarray]  # Weights-bias pair for a layer.
+ParamsList = list[LayerParam]
+
+def net(params: ParamsList, x: jnp.ndarray):
+  for W, b in params:
+    x = jnp.maximum(jnp.dot(x, W) + b, 0.)
+  return x
+```
+
+Instead, iterate over the layer application with {func}`jax.lax.scan`:
+
+```python
+StackedWeights = jnp.ndarray  # All weight matrices stacked together.
+StackedBiases = jnp.ndarray   # All bias vectors stacked together.
```

Comment:
I don't understand the reason for defining these aliases

Reply:
Same. This is *actually* self-explanatory.